### ISSUE TITLE
Prototype Pollution Vulnerability (by Sonatype CLM)

### ISSUE THREAD

**Issue by dvasconcelos1** [lodash/lodash#4738]
Follows the vulnerability report from Sonatype CLM:

> EXPLANATION
The lodash package is vulnerable to Prototype Pollution. The template function in lodash.js, template.js, and lodash.min.js does not account for unicode newline characters when filtering the sourceURL property of the options object. Because of how the options object is used, an attacker who can control the source URL can leverage this to alter properties on the prototype chain, which can cause other sections of code to behave in an arbitrary and malicious way.
>
> Please note that this vulnerability is due to an incomplete fix in sonatype-2019-0500.
>
> DETECTION
>The application is vulnerable by using this component.
>
>RECOMMENDATION
>There is no non vulnerable version of this component/package. We recommend investigating alternative components or a potential mitigating control.
>
>ROOT CAUSE
>lodash-4.17.15.tgzMETA-INF/resources/webjars/lodash/4.17.13/template.js[4.17.13, )
>ADVISORIES
>Project:https://github.com/lodash/lodash/pull/4517
>Project:https://github.com/lodash/lodash/pull/4518
>CVSS DETAILS
>Sonatype CVSS 3:9.8
>CVSS Vector:CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H

**Comment by dsiah** [comment]
Similarly I have found npm audit reporting numerous vulnerabilities. As of the time of this writting
```

> fsevents@1.2.9 install /Users/dsiah/OSS/lodash/node_modules/fsevents
> node install
...
added 753 packages from 919 contributors and audited 4103 packages in 23.792s
found 68 vulnerabilities (25 low, 19 moderate, 22 high, 2 critical)
  run `npm audit fix` to fix them, or `npm audit` for details
```
Do we have specific issues for this?

**Comment by mdeknowis** [comment]
Is that the same as reported https://hackerone.com/reports/712065 already 2019-07-19 and disclosed 2020-04-28 ?

**Comment by Kashish2394jn** [comment]
Yes , how soon can we expect a patch fix for this issue?

**Comment by mdeknowis** [comment]
Is this https://github.com/lodash/lodash/pull/4759 related?

**Comment by mdeknowis** [comment]
Also Whitesource considers it as a high vulnerability:
* CVSS 3 Score: 8.1
* CVSS 2 Score: 8.1

**Comment by mdeknowis** [comment]
It looks like, there is also another PR related to this: https://github.com/lodash/lodash/pull/4745

**Comment by mdeknowis** [comment]
The comment of https://github.com/lodash/lodash/pull/4745#issuecomment-622477124:

>  jdalton commented 11 days ago (May 1, 2020, 7:15 PM GMT+2)
> I'm aware of the issue and for me at least this is not a high priority issue. All disclosures and issues are not high priority and this one while technically an issue would require zipping objects based on user provided properties arrays . So yes, I will address this but I'm not letting some report ruin my week or plans.

**Comment by DevRCRun** [comment]
That's absolutely reasonable and the explanation of where the issue would be exploitable is really helpful, it just feels a little weird given the hackerone discussion is five months old

**Comment by orokanasaru** [comment]
That explanation is ridiculous. How does approving a PR ruin his week?
Ignoring communication with security companies until they are forced to publicly disclose a high sev vulnerability and then rejecting a fix and locking discussion is wildly unprofessional.

**Comment by AJHenry** [comment]
@orokanasaru 
I don't think you realize how hard it is to maintain one of the largest packages in existence, used by more than 6.9 MILLION direct dependents. Solving vulnerabilities is no trivial task and is much more involved than just accepting a PR. Just look at what happened to [request](https://github.com/request/request/issues/2772) when they published a bad version.

Besides, the attack surface is smaller than past vulnerabilities like `defaultsDeep` and such and you're _probably_ not affected by it (you should only import methods you need anyway, instead of the entire library).

 I think you can wait for a release for a vulnerability that doesn't even have a CVE. Just be patient.

**Comment by orokanasaru** [comment]
I'm not asking for anyone to drop everything to fix this. 

They had half a year between private and public disclosure.
The issue affects a small and well understood area.
They have a known pattern for fixing prototype pollution issues.
If there was any concern about risk, they could have communicated it and delayed public disclosure longer.
Instead, they ignored the issue for half a year hoping it would go away.

When someone from the community fixed it for them, they could have accepted it.
They could have given reasons for uncertainty about the fix and asked for more tests/investigations/etc.
They could have released a preview package for people to test in the wild.
They did not.
They whined about how how annoying PRs are, locked discussion, and once again hoped it would just go away.

Some of the 6.9 MILLION direct dependencies (and their transitive dependents) have legal and contractual requirements for how they handle reported vulnerabilities in their repositories.
Telling auditors that it's probably not a real issue because someone on lodash said so doesn't really work.

The unfortunate reality of getting popular is that you sometimes have to do things you don't like, including fixing bugs you don't personally care about and communicating about topics you don't want to talk about.

**Comment by taltal78** [comment]
Hi,
Is there a plan to fix this in the near future? 
or we must drop lodash usage?

thanks,
Tal

**Comment by DaveStein** [comment]
@jdalton @blikblum any update here?

**Comment by martinoppitz** [comment]
@orokanasaru :speech_balloon: best explanation, thanks!

https://github.com/lodash/lodash/issues/4738#issuecomment-629698149

**Comment by martinoppitz** [comment]
see https://github.com/lodash/lodash/issues/4786#issuecomment-633005670

**Comment by nickswope** [comment]
@orokanasaru is right. Here is another PR where we’ve been trying to get a sonotype version merged. That PR is 7 months old and I think that thread will provide others more context around what @orokanasaru has said

**Comment by EmanuilIvanovv** [comment]
Does anyone have a proper mitigation in place? I mean for direct dependency it's easy, but how can we be sure that baseZipObject is not used anywhere within all the other 3P libraries that use lodash?

**Comment by MarkDavidson** [comment]
The maintainer's response to this issue has caused my company to remove lodash from our products. I am not entitled to the maintainer's time or weekends, but I am responsible for delivering a secure product to paying customers.

There are far too many software vulnerabilities, and we all need to do our part to improve security. I hope the maintainer recognizes that he has significant leverage and can fix millions of vulnerabilities with one change.

**Comment by martinoppitz** [comment]
@MarkDavidson agree

**Comment by marty-wang** [comment]
The Tech community has never been short of virtue signalers. Where is the compassion that you virtue signalers have always been talking about?

**Comment by YuriGor** [comment]
![image](https://user-images.githubusercontent.com/6861317/83514152-5b988b80-a4d3-11ea-87a7-1e66b94fc50d.png)

Doesn't look like a Christmas :cry: 
I am wondering if v5 is about to be released and consuming all the time available?

**Comment by chrisVillanueva** [comment]
Hi Team.

I really enjoy lodash.

I'm going to begin another Enterprise project that must pass multiple security scans.   But like @MarkDavidson, we have to deliver secure code.  I'm not sure we'll use lodash this time around.

**Comment by rleibman** [comment]
I do not use javascript enough that I would even know what to do or where to start, but could it be forked and fixed? I'm just following because I'm in charge of keeping track of vulnerabilities in my team.

**Comment by martinoppitz** [comment]
@rleibman Stop using lodash. I think, today you can do without it.

See maintainer Motivation!

**Comment by rleibman** [comment]
> @rleibman Stop using lodash. I think, today you can do without it.

It's upstream from one of the things we use, I have only the barest of ideas of what it does, if I could, I would, get rid of a dumb Jira ticket.

**Comment by dwiyatci** [comment]
@dvasconcelos1 Did you mean: ~Vulneravility~ **_Vulnerability_**?

**Comment by Barkole** [comment]
* https://snyk.io/vuln/SNYK-JS-LODASH-567746
* https://snyk.io/test/npm/lodash/4.17.15

> **Prototype Pollution**
>  
>     Vulnerable module: lodash
>     Introduced through: lodash@4.17.15
> 
> **Detailed paths**
> 
>     Introduced through: lodash@4.17.15
>     Remediation: Open PR to patch lodash@4.17.15.
> 
> **Overview**
> 
> lodash is a modern JavaScript utility library delivering modularity, performance, & extras.
> 
> Affected versions of this package are vulnerable to Prototype Pollution. The function zipObjectDeep can be tricked into adding or modifying properties of the Object prototype. These properties will be present on all objects.
> PoC
> 
> ```js
> const _ = require('lodash');
> _.zipObjectDeep(['__proto__.z'],[123])
> console.log(z) // 123
> ```

**Comment by sjahn14** [comment]
It seems to have been fixed in 4.17.19 no ?

**Comment by kelset** [comment]
Yes, it seems so https://github.com/lodash/lodash/wiki/Changelog#v41712

**Comment by bnjmnt4n** [comment]
The bug has been fixed in v4.17.19.

**Comment by chrisVillanueva** [comment]
I will upgrade to v4.17.19 in one enterprise app.  This app currently surfaces a security violation for "prototype pollution vulnerability" in the `merge`, `mergeWith`, and `defaultsDeep` functions. I'll see if XRAY approves the update.
